src/DataTypes/convertMySQLDataType.cpp

```diff
@@ -25,48 +28,69 @@ ASTPtr dataTypeConvertToQuery(const DataTypePtr & data_type)
 {
+    // we expect mysql_data_type to be either "basic_type" or "type_with_params(param1, param2, ...)"
+    auto data_type = std::string_view(mysql_data_type);
+    const auto param_start_pos = data_type.find("(");
+    const auto type_name = data_type.substr(0, param_start_pos);
+
     DataTypePtr res;
-    if (mysql_data_type == "tinyint")
+    if (type_name == "tinyint")
     {
         if (is_unsigned)
             res = std::make_shared<DataTypeUInt8>();
         else
             res = std::make_shared<DataTypeInt8>();
     }
-    else if (mysql_data_type == "smallint")
+    else if (type_name == "smallint")
     {
         if (is_unsigned)
             res = std::make_shared<DataTypeUInt16>();
         else
             res = std::make_shared<DataTypeInt16>();
     }
-    else if (mysql_data_type == "int" || mysql_data_type == "mediumint")
+    else if (type_name == "int" || type_name == "mediumint")
     {
         if (is_unsigned)
             res = std::make_shared<DataTypeUInt32>();
         else
             res = std::make_shared<DataTypeInt32>();
     }
-    else if (mysql_data_type == "bigint")
+    else if (type_name == "bigint")
     {
         if (is_unsigned)
             res = std::make_shared<DataTypeUInt64>();
         else
             res = std::make_shared<DataTypeInt64>();
     }
-    else if (mysql_data_type == "float")
+    else if (type_name == "float")
         res = std::make_shared<DataTypeFloat32>();
-    else if (mysql_data_type == "double")
+    else if (type_name == "double")
         res = std::make_shared<DataTypeFloat64>();
-    else if (mysql_data_type == "date")
+    else if (type_name == "date")
         res = std::make_shared<DataTypeDate>();
-    else if (mysql_data_type == "datetime" || mysql_data_type == "timestamp")
+    if (type_name == "timestamp" && scale == 0)
+    {
         res = std::make_shared<DataTypeDateTime>();
-    else if (mysql_data_type == "binary")
+    }
+    else if (type_name == "datetime" || type_name == "timestamp")
+    {
+        res = std::make_shared<DataTypeDateTime64>(scale);
+    }
+    else if (type_name == "binary")
         res = std::make_shared<DataTypeFixedString>(length);
-    else
+    else if (type_name == "numeric" || type_name == "decimal")
+    {
+        if (precision <= DecimalUtils::maxPrecision<Decimal32>())
+            res = std::make_shared<DataTypeDecimal<Decimal32>>(precision, scale);
+        else if (precision <= DecimalUtils::maxPrecision<Decimal64>())
+            res = std::make_shared<DataTypeDecimal<Decimal64>>(precision, scale);
+        else if (precision <= DecimalUtils::maxPrecision<Decimal128>())
+            res = std::make_shared<DataTypeDecimal<Decimal128>>(precision, scale);
+    }
+
+    if (!res)
         /// Also String is fallback for all unknown types.
         res = std::make_shared<DataTypeString>();
```

Comment:
It's also a backward incompatible place. We need a setting with enum: which types are supported. There're also Decimal256 and Decimal512 that should be under this setting in future. Something like `supported_mysql_types = basic |  parameterized | extended_parameterized`.

Reply:
added a MultiEnum setting `mysql_datatypes_support_level` with possible values:
* decimal
* datetime64
* or empty (meaning old behaviour)/any combination of above